TFNetwork.py

```diff
@@ -779,16 +779,33 @@ def get_trainable_params(self):
     trainable_vars_col = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)
     assert isinstance(trainable_vars_col, list)
     l = []  # type: list[tf.Variable]
-    for layer_name in sorted(self._selected_train_layers):
-      layer = self.layers[layer_name]
-      assert isinstance(layer, LayerBase)
-      if not layer.trainable:
-        continue
-      for param_name, param in sorted(layer.params.items()):
-        assert isinstance(param, tf.Variable)
-        if param in trainable_vars_col:
-          l.append(param)
-          trainable_vars_col.remove(param)
+
+    def get_trainable(layer, l, trainable_vars_col):
+      if not isinstance(layer, LayerBase):
+        return l, trainable_vars_col
+      from TFNetworkRecLayer import RecLayer, _SubnetworkRecCell
+      if isinstance(layer, RecLayer) and isinstance(layer.cell, _SubnetworkRecCell) and layer.trainable:
+        sub_layers = list(layer.cell.net.layers.values())
+        if layer.cell.input_layers_net:
+          sub_layers += layer.cell.input_layers_net.layers.values()
+        if layer.cell.output_layers_net:
+          sub_layers += layer.cell.output_layers_net.layers.values()
+        for sub_layer in sub_layers:
+          l, trainable_vars_col = get_trainable(sub_layer, l, trainable_vars_col)
+      else:
+        if not layer.trainable:
+          return l, trainable_vars_col
+        for param_name, param in layer.params.items():
+          assert isinstance(param, tf.Variable)
+          if param in trainable_vars_col:
+            l.append(param)
+            trainable_vars_col.remove(param)
+      return l, trainable_vars_col
+
+    layers = [self.layers[layer_name] for layer_name in sorted(self._selected_train_layers)]
+    for layer in layers:
+      l, trainable_vars_col = get_trainable(layer, l, trainable_vars_col)
```

Comment:
This makes it way too complicated.
Why not check sth like `trainable = param in param.graph.get_collection_ref(tf.GraphKeys.TRAINABLE_VARIABLES)`? Or extend `LayerBase.add_param`?

Reply:
We cannot rely on `tf.GraphKeys.TRAINABLE_VARIABLES`, because even untrainable variables are added to this group. Eg. in the [LinearLayer](https://github.com/rwth-i6/returnn/blob/master/TFNetworkLayer.py#L2027), the matrix `W` is created without passing the `trainable` flag. Therefore, TF adds the `TRAINABLE_VARIABLES` graph key whether or not the layer should be trained.

I think this is the reason why the whole `get_trainable_params()` method is needed at all - we cannot just select all `TRAINABLE_VARIABLES`.
Maybe this is the main issue that should be fixed? It would require to update the parameter definition of every layer, but should remove the need for this method and the filtering. I do not know whether there will be any side effects of setting this flag.

Reply:
Of course we can do that. We could just change the `LayerBase.add_param` function.